### ISSUE TITLE
Draft message feature not working correctly

### ISSUE THREAD

**Issue by stephanieelliott** [Expensify/Expensify.cash#2100]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
Drafted message text should constrained to the user selected when the text was entered into the compose box, draft icon should show only when text exists in the compose box. 

## Actual Result: 
After drafting message text in the compose box for (but prior to sending), text continues to show after clicking into another chat, draft icon continues to show when text does not exist in the compose box. 

### Issue 1: 
**After drafting message text in the compose box for (but prior to sending), text continues to show after clicking into another chat (2 users)**

1. Open a chat with `User A` 
2. Enter text into the compose box (but do not send) 
3. Switch to chat with `User B` 
4. Notice: Message drafted in compose box to `User A` continues to show in compose box in chat with `User B` 

### Issue 2: 
**After drafting message text in the compose box for (but prior to sending), text continues to show after clicking into another chat (3 users)**
1. Open a chat with `User A` 
2. Enter text into the compose box (but do not send) 
3. Switch to chat with `User B` 
4. Switch to a chat with `User C`
5. Notice: Message drafted in compose box to `User A` continues to show in compose box in chat with `User C` 

### Issue 3: 
**Draft icon continues to show when text does not exist in the compose box** 
1. Open a chat with `User A` 
2. Enter text into the compose box (but do not send), wait 3-5 seconds 
3. Switch to chat with `User B` 
4. Go back to chat with `User A`, delete the text drafted in compose box 
5. Switch to chat with `User C` 
6. Notice: Draft icon continues to show next to `User A` even though you deleted the text in the compose box

## Workaround:
N/A

## Platform:
Desktop app v1.0.6-0 

**Notes/Photos/Videos:** 

**Issue 1: After drafting message text in the compose box for (but prior to sending), text continues to show after clicking into another chat (2 users)****

https://user-images.githubusercontent.com/31225194/112590102-5aa85180-8da6-11eb-90d4-b3cfe1f1d988.mp4

**Issue 2: After drafting message text in the compose box for (but prior to sending), text continues to show after clicking into another chat (3 users)**

https://user-images.githubusercontent.com/31225194/112589293-f0db7800-8da4-11eb-935a-48ff721c777a.mp4

------
**
**Draft icon continues to show when text does not exist in the compose box** 

https://user-images.githubusercontent.com/31225194/112589862-ed94bc00-8da5-11eb-81b1-d5ce14352df3.mp4

**Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/157277

**Comment by stephanieelliott** [comment]
View this job on Upwork: https://www.upwork.com/jobs/~0129016d527b7dd243

**Comment by thierrysantos** [comment]
I'm looking for the codebase and I saw that the we have a problem of "synchronization" with the Onyx state.
 
We have a value in the state of Onyx and other in the `TextInputFocusable`.

render () reportId: 71623989 with value: thierry
render () reportId: 71623940 with value: concierge

So the idea is to fix this difference between the states.

Proposal in upwork: https://www.upwork.com/ab/proposals/1375448538180206593

@iwiznia @stephanieelliott 💯

**Comment by kaushiktd** [comment]
Hi,

This is Kaushik here and this is my technical explanation for this issue.

1.  when the user types a text message to send but actually not press to send and choose another person, in the meantime, we need to clear the text input value and remain this value for the user whom we want to send.
2.  when work on the first issue second will auto resolved.
3.  we need to check in existing code if the draft message is existing or not, if not then remove the draft icon.

**Comment by iwiznia** [comment]
@thierrysantos how exactly would you `fix this difference between the states`?

@kaushiktd can you add more details to your proposal, like where in the code you would make the changes and what would those changes be?

**Comment by thierrysantos** [comment]
@iwiznia 

Just to give a feedback for you, I know the problem but I'm thinking in a solution. 🤔

But basically the `TextInputFocusable` is an uncontrolled input, so when we change the context (reportID) this component have their own value and don't receive the new value from the switch between reportID.

And the change doesn't occur because we sent the value in the `defaultValue` property and when we change the context(reportID) this input alredy have a value, so `defaultValue` isn't used as a initial value to `value` property.

That's okay you all hiring other developer but I will continue here to think in how to fix it.

**Comment by thierrysantos** [comment]
@iwiznia I have find a solutions to this problem(`TextInput` and `draft icon`).

Basically when change the context, we could re-render the component(`ReportActionCompose`) and it will solve this problems.

**Comment by iwiznia** [comment]
> (with key property)

To clarify, what would the key be? The reportID you are looking at?

> Because it will use the value from onyx as defaultValue in TextInputFocusable.

Which value exactly? The one saved in `${ONYXKEYS.COLLECTION.REPORT_DRAFT_COMMENT}${reportID}`?

**Comment by thierrysantos** [comment]
Would be like `<ReportActionCompose key={this.props.reportID} />` this value come from `ONYXKEYS.CURRENTLY_VIEWED_REPORTID`.

**Comment by iwiznia** [comment]
Cool, that's what I thought. How about the other question?
> Which value exactly? The one saved in ${ONYXKEYS.COLLECTION.REPORT_DRAFT_COMMENT}${reportID}?

**Comment by thierrysantos** [comment]
> Which value exactly? The one saved in ${ONYXKEYS.COLLECTION.REPORT_DRAFT_COMMENT}${reportID}?

No, the value saved in `${ONYXKEYS.CURRENTLY_VIEWED_REPORTID}`.

**Comment by iwiznia** [comment]
Huh? The value of the text input would be the reportID? I don't think that's correct, should be the draft comment of that reportID (if there's any)

**Comment by thierrysantos** [comment]
Oh sorry, I made a mistake here, the value is the draft from the comment of that reportID.

> The one saved in ${ONYXKEYS.COLLECTION.REPORT_DRAFT_COMMENT}${reportID}?

Yeah

**Comment by iwiznia** [comment]
Cool, that makes sense! Go ahead and submit the PR.

**Comment by thierrysantos** [comment]
I sent a proposal on upwork, do I need some one to accept the proposal or I could start right now?

**Comment by thierrysantos** [comment]
I sent a email to join in your slack channel

**Comment by Jag96** [comment]
@thierrysantos offer sent!

**Comment by kidroca** [comment]
Hey guys since you will make the compose component re-mount you can also set the `autoFocus` attribute which would make it focus after it mounts. 

```jsx
<TextInputFocusable
  autoFocus
  multiline
  ref={el => this.textInput = el}
  ...
```

Then this condition becomes unneeded - when reportID changes the compose component would remount:

https://github.com/Expensify/Expensify.cash/blob/62b6920c01cdefbc0fb0c258e4b756bcd01fe12a/src/pages/home/report/ReportActionCompose.js#L84

```js
        if (
            (prevProps.modal.isVisible && !this.props.modal.isVisible)
            || (prevProps.reportID !== this.props.reportID) // this won't be needed 
        ) {
            this.setIsFocused(true);
        }
```

I've proposed a similar idea though it was regarding focusing and keyboard issues on mobile 

In short:
The `ReportView` (or screen) will re-mount. This in turn means it's children will remount too and you won't need to set a `key` to force a remount

- Why does the `ReportView` need to unmount?
- To have a smooth transition while we are switching between reports  #2154 

This can simplify logic like:

```js
  // The first time the component loads the props is empty and the next time it may contain value.
  // If it does let's update this.comment so that it matches the defaultValue that we show in textInput.
  if (this.props.comment && prevProps.comment === '' && prevProps.comment !== this.props.comment) {
      this.comment = this.props.comment;
  }
```

The parent will first load data (display loading state) and then mount the compose field with a proper initial value.

**Comment by aliabbasmalik8** [comment]
Hi.
The simple solution to this issue is to add `Key` like this way. 

![image](https://user-images.githubusercontent.com/31027036/113226367-e8dd7580-9287-11eb-935f-67de2d738944.png)


https://user-images.githubusercontent.com/31027036/113226523-57223800-9288-11eb-87d6-de4045cc0a5e.mov

**Comment by thierrysantos** [comment]
@iwiznia what is your JDK version and node version? (I'm with trouble to run the app in android)

**Comment by thierrysantos** [comment]
The PR is open guys https://github.com/Expensify/Expensify.cash/pull/2166
